### PR TITLE
648 move pg conversion to imputation.

### PR DESCRIPTION
**Pull request by AnneONS**
### Note: the code runs, and tests work, but I  haven't checked the outputs

PG conversion now happens at the beginning of the imputation module. The PG column, 201, is initially numeric, and contains many nulls.

This means that in construction, col 201 can remain numeric as in spp.

Step 1 of TMI is no longer needed.

The approach to PG conversion is now:

1.  Fill all nulls in column 201 with numeric pg values from the sic mapper
2. Copy this column to a new column, "pg_numeric"
3. Convert column 201 to alpha-numeric

The pg_numeric column will now have nulls filled for apportionment, and for the tau and sas outputs.

In outputs, the pg conversion is done for the NI data.

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/imputation/pg_conversion.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    # Flag all SIC numbers that don't have a corresponding map value
+    mapless_errors = []
+    for key, value in map_dict.items():
+        if str(value) == "nan":
+            mapless_errors.append(key)
+
+    if mapless_errors:
+        PgLogger.error(
+            f"Mapping doesnt exist for the following SIC numbers: {mapless_errors}"
+        )
```
Comment (coatet) [inline src/imputation/pg_conversion.py:56]: Is the logged error just to alert users or should this raise an exception? Not sure what kind of scheme you decided on for when to use error vs warning logging events.

File: src/imputation/pg_conversion.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    df: pd.DataFrame,
+    mapper: pd.DataFrame,
+    pg_column: str = "201",
+    from_col: str = "pg_numeric",
+    to_col: str = "pg_alpha",
+):
+    """Map from PG numeric to PG alpha-numeric and create a new column.
+
+    The product group column (default: column 201) coped to a new column, "pg_numeric",
+    and then is updated from numeric to alpha-numeric using a mapping.
```
Comment (coatet) [inline src/imputation/pg_conversion.py]: Should be "is copied"?

File: src/imputation/pg_conversion.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    # Flag all PGs that don't have a corresponding map value
+    mapless_errors = []
+    for key, value in map_dict.items():
+        if str(value) == "nan":
+            mapless_errors.append(key)
+
+    if mapless_errors:
+        PgLogger.error(
+            f"Mapping doesnt exist for the following product groups: {mapless_errors}"
+        )
```
Comment (coatet) [inline src/imputation/pg_conversion.py:126]: As above, just wanted to check whether this should raise an exception or not.

File: src/imputation/pg_conversion.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        to_col (str, optional): The column in the mapper that is used to map to.
+
+    Returns:
+        pd.DataFrame: A dataframe with all target column values mapped
+    """
+
+    df = df.copy()
+
+    # Create a mapping dictionary from the 2 columns
+    map_dict = dict(zip(sicmapper[from_col], sicmapper[to_col]))
```
Comment (coatet) [inline src/imputation/pg_conversion.py:46]: I get a KeyError for  `SIC 2007_CODE` here

**Review by coatet — CHANGES_REQUESTED** [review]
All the code looks good on a readthrough but I'm getting some errors when running. Can't seem to highlight specific lines in some files so I'm putting them here. 

Line 11 of `staging_main`:
```python
from src.staging import pg_conversion as pg
```
This fails since it's been moved to imputation. 

Line 289 of `staging_main`: 
```python
 # Map PG from SIC/PG numbers to column '201'.
full_responses = pg.run_pg_conversion(full_responses, pg_num_alpha, sic_pg_alpha_mapper, target_col="201")
```
This fails with `got an unexpected keyword argument 'target_col'` since it looks like the argument should be `pg_column` instead of `target_col`. 

When I fixed these I got it to run as far as Line 46 in `pg_conversion` but that prompts what looks like an error in columns, not sure I know how to fix that.

File: src/imputation/pg_conversion.py (hunk 2 above)
Comment (AnneONS) [reply src/imputation/pg_conversion.py]: yes ha ha

**Review by coatet — APPROVED** [review]

**Pull request merged**